src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -178,10 +180,19 @@ private void CheckAndResizeBuffer(int sizeHint)
 
                 if ((uint)newSize > int.MaxValue)
                 {
-                    newSize = currentLength + sizeHint;
+                    // Attempt to grow by the larger of the minimum size and half of the available size.
```

Comment:
Instead of doing this, we could allocate the magic number that represents the largest array size, but allocating half of the available size seems less risky.

Reply:
> the magic number that represents the largest array size

It is what we do in number of other places. I do not see a problem with doing it here as well.

We can consider adding API that returns the max size to make it less magic.

Reply:
What would be that magic number? `int.max - 56`, or `2 billion`, or something else?

Until we have such an API, let's add it as a named constant here, similar to `DefaultInitialBufferSize = 256`.

Either approach would be reasonable.

The downside of going all the way to a ~2 billion right away, is that for output data that is around 1.1 to 1.5 billion, we are over-allocating by quite a lot. So, the current approach if incrementing by half-way between current and max, has some benefit there.

On the other hand, the closer to you get to 2 billion by incrementing the size this way, that increasing exponential decay will become small again (possible even 2-100 bytes).

Reply:
> What would be that magic number? int.max - 56, or 2 billion, or something else?

It varies a bit between byte and other numbers : https://github.com/dotnet/runtime/search?q=MaxArrayLength

For now, I think using the max makes sense since we already allocated size of at least `int32.MaxValue / 2` so we would always allocate less than double the current buffer size; i.e. the current algorithm doubles on each alloc, so the last alloc using max doesn't quite double so it seems fine from that viewpoint.

Reply:
> What would be that magic number? int.max - 56, or 2 billion, or something else?

https://github.com/dotnet/runtime/blob/85162fbf9ccdeb4fa1df357f27308ae96579c066/src/libraries/System.Private.CoreLib/src/System/Array.cs#L21-L27